### PR TITLE
Browse links and "by regex" links in item display (REST)

### PR DESCRIPTION
**Pull request by kshepherd**
## References
_Add references/links to any related issues or PRs. These may include:_
* Related to [Angular PR#1972](https://github.com/DSpace/dspace-angular/pull/1972)
* Related to [REST Contract#205](https://github.com/DSpace/RestContract/pull/205)

## Description
This change allows the following configurable display of links in item page metdata values:
1. The configuration in `webui.browse.links.<n>` (dspace.cfg) now works as it did in previous versions, where a browse link is rendered for the metadata value in question if it is configured for a valid browse index. This works for both plain item page field components, and for metadata representations where there is virtual metadata instead of a link to a related entity. (related entity link will always take precedence)
3. The item page field components (and children) now allow a `[urlRegex]` declarative in the templates where a link will be rendered if and only if the regex matches the metadata text value, eg. `^https?://`, or something more restrictive

## To test conditional regex-based linking:

Try adding a section to a typed or untyped item template with a field like:
```
<ds-generic-item-page-field [item]="object"
      [fields]="['dc.subject']"
      [separator]="', '"
      [label]="'item.page.subject'">
      [urlRegex]="'^https?://www.something.com'"
    </ds-generic-item-page-field>
```

And try it out with different subject values, some of which match the regular expression.

## To test browse links:

Set up browse link config for fields, such as:
```
webui.browse.link.1 = author:dc.contributor.*
webui.browse.link.2 = subject:dc.subject.*
```

And try out items with values in these fields. With the author field, note that if it is not a "authorOf..." relation, the browse author link should be displayed. If it *is* a relation to a person entity, it should instead be linked to that entity.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.

### CODE REVIEW

**Review by tdonohue — CHANGES_REQUESTED** [review]
@kshepherd : I wanted to make sure you saw the feedback on the REST Contract regarding the new `/api/discover/browselinks` endpoint: https://github.com/DSpace/RestContract/pull/205#issuecomment-1379337562

I think this PR (and the corresponding Angular PR) would need to be refactored before we can move it forward.  We've made some suggestions on a possible new endpoint design in the REST Contract comment above.

**Comment by kshepherd** [comment]
@tdonohue I have refactored this as requested on the Rest Contract note. The /browses/ endpoint, controller and data services are now used. I have fixed integration tests.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@kshepherd : Gave this a test today and overall it *works* well.  However, I noticed there's a few endpoints here that don't appear to be used at all (by the UI) and I think they can simply be removed (as it vastly simplifies both this PR and the UI PR).

Also, I'd recommend we remove this "UNSUPPORTED in DSpace 7.0" note in dspace.cfg in this PR: https://github.com/DSpace/DSpace/blob/main/dspace/config/dspace.cfg#L1232

A few more inline comments below.  Overall this looks good and I'm basically a +1 once this is cleaned up.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java [hunk 1]
```diff
@@ -1487,4 +1487,74 @@ public void testBrowseByDateIssuedItemsFullProjectionTest() throws Exception {
...
+                //The array of browse index should have a size 1
+                .andExpect(jsonPath("$.id", is("author")))
+
+                //Check that all (and only) the default browse indexes are present
+                .andExpect(jsonPath("$.metadataBrowse", is(true)))
+        ;
+    }
+
+    @Test
+    public void findOneLinkedPassingTwoFields() throws Exception {
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java:1512]: I have added two tests - ensuring that a search for field(s) with no browse config returns a 204 No Content, and that if the fields parameter is missing, we get a 400 Bad Request

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java [hunk 2]
```diff
@@ -1487,4 +1487,74 @@ public void testBrowseByDateIssuedItemsFullProjectionTest() throws Exception {
...
+    }
+
+    /**
+     * Expect a list of browse definitions that are also configured for link rendering
+     * @throws Exception
+     */
+    @Test
+    public void findAllLinked() throws Exception {
+        //When we call the root endpoint
+        getClient().perform(get("/api/discover/browses/search/allLinked"))
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java]: Done

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java [hunk 3]
```diff
@@ -1487,4 +1487,74 @@ public void testBrowseByDateIssuedItemsFullProjectionTest() throws Exception {
...
     }
+
+    /**
+     * Expect a single author browse definition
+     * @throws Exception
+     */
+    @Test
+    public void findOneLinked() throws Exception {
+        //When we call the root endpoint
+        getClient().perform(get("/api/discover/browses/search/byField?field=dc.contributor.author"))
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/BrowsesResourceControllerIT.java]: Done

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/BrowseIndexRestRepository.java [hunk 4]
```diff
@@ -53,6 +59,96 @@ public Page<BrowseIndexRest> findAll(Context context, Pageable pageable) {
...
+     * Get paginated list of all browse index definitions for configured browse links
+     *
+     * @param context
+     *            the dspace context
+     * @param pageable
+     *            object embedding the requested pagination info
+     * @return
+     */
+    @SearchRestMethod(name = "allLinked")
+    public Page<BrowseIndexRest> findAllLinked(Context context, Pageable pageable) {
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/BrowseIndexRestRepository.java]: Done

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/BrowseIndexRestRepository.java [hunk 5]
```diff
@@ -53,6 +59,96 @@ public Page<BrowseIndexRest> findAll(Context context, Pageable pageable) {
...
     }
 
+    /**
+     * Find a browse index by a specific field
+     * @param field
+     * @return
+     * @throws SQLException
+     */
+    @SearchRestMethod(name = "byField")
+    public BrowseIndexRest findByField(@Parameter(value = "field", required = true) String field)
```
[the remark this answers is not in the document]
Comment (kshepherd) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/BrowseIndexRestRepository.java]: Done

**Comment by kshepherd** [comment]
@tdonohue I believe I've addressed all the review feedback, thanks.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @kshepherd !  This looks good to me now, and aligns with the REST Contract.

**Comment by tdonohue** [comment]
Merging with 1 approval as this is small and aligns with the contract (which has had several reviewers & all feedback has been addressed).

**Pull request merged**